vault/logical_system_events.go

```diff
@@ -5,26 +5,209 @@ package vault
+		b.logger.Error("Error initializing plugin", "error", err)
+		return logical.ErrorResponse("error initializing plugin"), logical.ErrUnrecoverable
+	}
+	if err != nil {
+		b.logger.Warn("Error subscribing to events", "error", err)
+		return logical.ErrorResponse("error subscribing to events"), logical.ErrUnrecoverable
+	}
+	ns, err := namespace.FromContext(requestCtx)
+	if err != nil {
+		b.logger.Error("No namespace found", "error", err)
+		return logical.ErrorResponse("no namespace found"), logical.ErrUnrecoverable
+	}
+	ctx, ctxCancel := context.WithCancel(context.Background())
+	eventCh, eventCancel, err := b.Core.events.Subscribe(ctx, ns, eventTypePattern, bexprFilter)
+	sub := &eventSubscription{
+		ctx:     ctx,
+		id:      id,
+		plugin:  plugin,
+		config:  config,
+		backend: backend,
+	}
+	sub.cancel = func() {
+		sub.cancel = func() {} // for safety
+		ctxCancel()
+		eventCancel()
+	}
+	err = sub.backend.Subscribe(requestCtx, &event.SubscribeRequest{
+		SubscriptionID:   id,
+		Config:           config,
+		VerifyConnection: false, // TODO: read this from parameters
+	})
+	if err != nil {
+		b.logger.Warn("Error starting subscription", "error", err)
+		return logical.ErrorResponse(fmt.Sprintf("error starting subscription: %v", err)), logical.ErrUnrecoverable
+	}
+	b.eventSubscriptions.subscriptions.Put(id, sub)
+	go b.subscriptionLoop(sub, eventCh)
+	resp := &logical.Response{
+		Data: map[string]interface{}{
+			"id":     id,
+			"plugin": plugin,
+		},
+	}
+	return logical.RespondWithStatusCode(resp, req, http.StatusOK)
+}
+
+func (b *SystemBackend) unsubscribe(sub *eventSubscription) {
+	sub.cancel()
+	go func() {
+		// subscription context might have been canceled already
+		err := sub.backend.Unsubscribe(context.Background(), &event.UnsubscribeRequest{
+			SubscriptionID: sub.id,
+		})
+		if err != nil {
+			b.logger.Warn("Error unsubscribing", "subscription_id", sub.id, "error", err)
+		}
+	}()
+}
+
+func (b *SystemBackend) subscriptionLoop(sub *eventSubscription, eventCh <-chan *eventlogger.Event) {
+	doneCh := sub.ctx.Done()
+	defer b.unsubscribe(sub)
+	for {
+		select {
+		case <-doneCh:
+			return
+		case payload := <-eventCh:
+			jsonBytes, ok := payload.Format("cloudevents-json")
+			if !ok {
+				b.logger.Error("Event system is not supplying 'cloudevents-json' format")
+				return
+			}
+			err := sub.backend.Send(sub.ctx, &event.SendRequest{
+				SubscriptionID: sub.id,
+				EventJSON:      string(jsonBytes),
+			})
+			if err != nil {
+				b.logger.Warn("Error sending event", "subscription_id", sub.id, "error", err)
+				return
+			}
+		}
+	}
 }
 
 // handleEventsUnsubscribe
-func (b *SystemBackend) handleEventsUnsubscribe(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
-	// TODO
+func (b *SystemBackend) handleEventsUnsubscribe(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	plugin := data.Get("plugin").(string)
```

Comment:
may want to handle the type assert here. Not sure if `GetOk()` is a good fit here.